java/src/main/native/src/row_conversion.cu

```diff
@@ -1670,6 +1737,76 @@ std::vector<std::unique_ptr<column>> convert_to_rows(table_view const &tbl,
+  auto const num_columns = tbl.num_columns();
+  auto const num_rows = tbl.num_rows();
+
+  auto const fixed_width_only = std::all_of(
+      tbl.begin(), tbl.end(), [](column_view const &c) { return is_fixed_width(c.type()); });
+
+  // break up the work into tiles, which are a starting and ending row/col #.
+  // this tile size is calculated based on the shared memory size available
+  // we want a single tile to fill up the entire shared memory space available
+  // for the transpose-like conversion.
+
+  // There are two different processes going on here. The GPU conversion of the data
+  // and the writing of the data into the list of byte columns that are a maximum of
+  // 2 gigs each due to offset maximum size. The GPU conversion portion has to understand
+  // this limitation because the column must own the data inside and as a result it must be
+  // a distinct allocation for that column. Copying the data into these final buffers would
+  // be prohibitively expensive, so care is taken to ensure the GPU writes to the proper buffer.
+  // The tiles are broken at the boundaries of specific rows based on the row sizes up
+  // to that point. These are row batches and they are decided first before building the
+  // tiles so the tiles can be properly cut around them.
+
+  std::vector<size_type> column_sizes;  // byte size of each column
+  std::vector<size_type> column_starts; // offset of column inside a row including alignment
+  column_sizes.reserve(num_columns);
+  column_starts.reserve(num_columns + 1); // we add a final offset for validity data start
+
+  auto schema_column_iter =
+      thrust::make_transform_iterator(thrust::make_counting_iterator(0),
+                                      [&tbl](auto i) -> std::tuple<data_type, column_view const> {
+                                        return {tbl.column(i).type(), tbl.column(i)};
+                                      });
+
+  auto const fixed_width_size_per_row = detail::compute_column_information(
+      schema_column_iter, schema_column_iter + num_columns, column_starts, column_sizes);
+  if (fixed_width_only) {
+    // total encoded row size. This includes fixed-width data and validity only. It does not include
+    // variable-width data since it isn't copied with the fixed-width and validity kernel.
+    auto row_size_iter = thrust::make_constant_iterator<uint64_t>(
+        util::round_up_unsafe(fixed_width_size_per_row, JCUDF_ROW_ALIGNMENT));
+
+    auto batch_info = detail::build_batches(num_rows, row_size_iter, fixed_width_only, stream, mr);
+
+    detail::fixed_width_row_offset_functor offset_functor(
+        util::round_up_unsafe(fixed_width_size_per_row, JCUDF_ROW_ALIGNMENT));
+
+    return detail::convert_to_rows(tbl, batch_info, offset_functor, column_starts, column_sizes,
+                                   fixed_width_size_per_row, stream, mr);
+  } else {
+    auto row_sizes = detail::build_string_row_sizes(tbl, fixed_width_size_per_row, stream);
```

Comment:
Would you be averse to combining the two branches into one? We might achieve this with a helper function template, with `T = decltype(row_size_iter)`.

Reply:
(What? We were all thinking it. 😛)

Reply:
It's possible that this might not be an improvement. Please feel free to reject.
My thought was to use the following helpers:
```c++
template <bool is_fixed_width = true>
fixed_width_row_offset_functor get_offset_functor(batch_data const& /*ignored*/, size_type fixed_width_size_per_row)
{
  return fixed_width_row_offset_functor{util::round_up_unsafe(fixed_width_size_per_row, JCUDF_ROW_ALIGNMENT)};
}

template <>
string_row_offset_functor get_offset_functor<false>(batch_data const& batch_info, size_type /*ignored*/)
{
  return detail::string_row_offset_functor{batch_info.batch_row_offsets};
}

template <bool fixed_width_only, typename RowSizeIter, typename OffsetsFunctor>
std::vector<std::unique_ptr<column>> convert_to_rows_impl(table_view const& tbl,
                                                          size_type const fixed_width_size_per_row,
                                                          RowSizeIter row_size_iter,
                                                          host_span<size_type> column_starts,
                                                          host_span<size_type> column_sizes,
                                                          rmm::cuda_stream_view stream,
                                                          rmm::mr::device_memory_resource *mr)
{
  auto const batch_info = detail::build_batches(tbl.num_rows(), row_size_iter, fixed_width_only, stream, mr);
  auto const offset_functor = get_offset_functor<fixed_width_only>(batch_info, fixed_width_size_per_row);
  return convert_to_rows(
    tbl, batch_info, offset_functor, column_starts, column_sizes, fixed_width_only, fixed_width_size_per_row, stream, mr);
}
```
This might enable us to marginally abbreviate the body of `convert_to_rows()` thus:
```c++
  if (fixed_width_only) {
    // total encoded row size. This includes fixed-width data and validity only. It does not include
    // variable-width data since it isn't copied with the fixed-width and validity kernel.
    auto const rounded_up_size_per_row = util::round_up_unsafe(fixed_width_size_per_row, JCUDF_ROW_ALIGNMENT);
    return detail::convert_to_rows_impl<true>(tbl, fixed_width_size_per_row, 
                                              thrust::make_constant_iterator<uint64_t>(rounded_up_size_per_row),
                                              column_starts, column_sizes, stream, mr);
  } else {
    auto row_sizes = detail::build_string_row_sizes(tbl, fixed_width_size_per_row, stream);
    auto row_size_iter = cudf::detail::make_counting_transform_iterator(
        0, detail::row_size_functor(num_rows, row_sizes.data(), 0));
    return detail::convert_to_rows_impl<false>(tbl, fixed_width_size_per_row, row_size_iter, column_starts,
                                               column_sizes, stream, mr);
  }

```

Reply:
While this removes some repetition in the code, I'm not certain it necessarily makes the code better.
@nvdbaranec, what say you?